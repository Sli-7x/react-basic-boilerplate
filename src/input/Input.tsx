import * as React from 'react';
import { StyledInput, ErrorText } from './Input.style';

export interface IInputProps {
  onChange: (evenr: Event) => void;
  value: string | number;
  label?: string;
  placeholder?: string;
  inputStyle?: { [key: string]: any };
  maxLength?: number;
  minLength?: number;
  errorText?: any;
  type?: string;
}

export class Input extends React.Component<IInputProps, {}> {
  public static defaultProps = {
    type: 'text',
  };

  constructor(props: IInputProps) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }

  // React.ChangeEvent<Event>
  handleChange(event: any) {
    this.props.onChange(event);
  }

  getErrors(errors: string | string[]) {
    if (errors) {
      if (Array.isArray(errors) && errors.length > 0) {
        return errors.map((val: string, i: number) => <ErrorText key={i}>{val}</ErrorText>);
      }
      if (typeof errors === 'string') {
        return <ErrorText>{errors}</ErrorText>;
      }
    }
  }

  render() {
    const { value, label, placeholder, inputStyle, maxLength, minLength, type, errorText } = this.props;

    const labelComp = label ? <label>{label}</label> : '';

    return (
      <div>
        {labelComp}
        <StyledInput
          style={inputStyle ? { ...inputStyle } : {}}
          type={type}
          value={value}
          onChange={this.handleChange}
          placeholder={placeholder}
          maxLength={maxLength ? maxLength : 255}
          minLength={minLength ? minLength : 0}
          isInvalid={errorText && errorText.length}
        />
        {this.getErrors(errorText)}
      </div>
    );
  }
}
