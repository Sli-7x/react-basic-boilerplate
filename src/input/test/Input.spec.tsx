import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import { Input, IInputProps } from '../Input';

describe('Input', () => {
  const onChange = jest.fn();
  const props: IInputProps = {
    onChange,
    value: '',
    inputStyle: {},
    label: 'label',
    maxLength: 20,
    minLength: 1,
  };

  it('should render correctly', () => {
    const tree = renderer.create(<Input {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('onChange()', () => {
    it('should call onChange', () => {
      const wrapper = mount(<Input onChange={onChange} value="" />);
      const input = wrapper.find('input');
      input.simulate('change', { target: { value: 'test' } });
      expect(onChange).toHaveBeenCalled();
    });
  });

  describe('getErrors()', () => {
    it('should cover when string passed', () => {
      const wrapper = shallow(<Input {...props} />);
      const instance = wrapper.instance() as Input;
      const errors = instance.getErrors('test');
      expect(errors).not.toBe(null);
    });

    it('should cover when array passed', () => {
      const wrapper = shallow(<Input {...props} />);
      const instance = wrapper.instance() as Input;
      const errors = instance.getErrors(['test']);
      expect(errors).not.toBe(null);
    });

    it('should do nothing if empty array passed', () => {
      const wrapper = shallow(<Input {...props} />);
      const instance = wrapper.instance() as Input;
      const errors = instance.getErrors([]);
      expect(errors).toBe(undefined);
    });

    it('should do nothing if empty array passed', () => {
      const wrapper = shallow(<Input {...props} errorText={'test'} />);
      const instance = wrapper.instance() as Input;
      expect(instance.props.errorText).toBe('test');
    });
  });
});
