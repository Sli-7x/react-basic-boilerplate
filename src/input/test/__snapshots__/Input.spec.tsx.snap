// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<div>
  <label>
    label
  </label>
  <input
    className="sc-bdVaJa cmfyDO"
    maxLength={20}
    minLength={1}
    onChange={[Function]}
    style={Object {}}
    type="text"
    value=""
  />
</div>
`;
